### PR TITLE
Add SNAT random-fully flag.

### PR DESCRIPTION
**Pull request by fasaxc**
## Description

Detect iptables version and pass feature flags around.  

- SNAT --random-fully was aded in iptables v1.6.1
- MASQUERADE --random-fully was added in iptables v1.6.2

Use feature flag to control whether the --random-fully flag is included.

Fixes https://github.com/projectcalico/calico/issues/2073

## Todos
- [x] Check kernel compat?
- [x] Unit tests (full coverage)
- [x] Integration tests (delete as appropriate) In plan/Not needed/Done
- [ ] Documentation
- [ ] Backport
- [ ] Release note

## Release Note

```release-note
Where supported, Felix now uses the iptables --random-fully option when programming SNAT and MASQUERADE rules.
```

### CODE REVIEW

File: iptables/feature_detect.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	featuresT := reflect.TypeOf(merged)
+	for i := 0; i < featuresT.NumField(); i++ {
+		if reflect.ValueOf(*a).Field(i).Bool() && reflect.ValueOf(*b).Field(i).Bool() {
+			reflect.ValueOf(&merged).Elem().Field(i).SetBool(true)
+		}
+	}
+	return &merged
+}
+
+// VersionToFeatures convers an iptables version line as returned by iptables --version into a set of feature flags.
```
Comment (neiljerram) [inline iptables/feature_detect.go]: *converts

File: iptables/table_test.go [hunk 2]
```diff
@@ -731,7 +754,7 @@ func describeDirtyDataplaneTests(appendMode bool) {
 		It("with no errors, it should get to correct final state", func() {
 			table.Apply()
 			checkFinalState()
-			Expect(len(dataplane.Cmds)).To(Equal(2)) // a save and a restore
+			Expect(dataplane.Cmds).To(HaveLen(3)) // a save and a restore
```
Comment (neiljerram) [inline iptables/table_test.go]: need comment update

File: iptables/table_test.go [hunk 3]
```diff
@@ -742,7 +765,7 @@ func describeDirtyDataplaneTests(appendMode bool) {
 				checkFinalState()
 			})
 			It("it should retry once", func() {
-				Expect(len(dataplane.Cmds)).To(Equal(3)) // 2 saves and a restore
+				Expect(dataplane.Cmds).To(HaveLen(4)) // 2 saves and a restore
```
Comment (neiljerram) [inline iptables/table_test.go]: here too

File: iptables/utils_for_test.go [hunk 4]
```diff
@@ -416,6 +434,65 @@ func (d *saveCmd) StdoutPipe() (io.ReadCloser, error) {
...
+func (d *saveCmd) Run() error {
+	return errors.New("Not implemented")
+}
+
+type versionCmd struct {
+	Dataplane *mockDataplane
+}
+
+func (d *versionCmd) String() string {
+	return "saveCmd"
```
Comment (neiljerram) [inline iptables/utils_for_test.go]: Is that right?  (As opposed to `"versionCmd"`)

**Comment by fasaxc** [comment]
I went that direction to make it easier to test in  fully-deterministic way.  A global variable doesn't fit well with ginkgo tests at all and once the table needs to be passed it, it needs to be passed down to the rules too (since the rules shouldn't be carrying around a feature flag pointer).

**Pull request merged**